cmd/kubeapps-apis/plugins/kapp_controller/packages/v1alpha1/server_data_resources.go

```diff
@@ -399,3 +406,117 @@ func (s *Server) updatePkgInstall(ctx context.Context, cluster, namespace string
+	typedClient, _, _, err := s.GetClients(ctx, cluster)
+	if err != nil {
+		return "", status.Errorf(codes.Internal, "unable to get the k8s client: '%v'", err)
+	}
+
+	// the ConfigMap name is, by convention, "<appname>-ctrl", but it will change in the near future
+	cmName := fmt.Sprintf("%s-ctrl", installedPackageRefId)
+	cm, err := typedClient.CoreV1().ConfigMaps(namespace).Get(ctx, cmName, metav1.GetOptions{})
+
+	if err != nil || cm.Data["spec"] != "" {
+		log.Warning(errorByStatus("get", "ConfigMap", cmName, err))
+	}
+
+	appLabelValue := extractValueFromJson(cm.Data["spec"], "labelValue")
+
+	return appLabelValue, nil
+}
+
+// findMatchingK8sResources returns the list of k8s resources matching the given listOptions
+// Code inspired by https://github.com/kubernetes/kubectl/blob/release-1.22/pkg/cmd/apiresources/apiresources.go#L142
+func (s *Server) findMatchingK8sResources(ctx context.Context, cluster string, listOptions metav1.ListOptions) ([]*corev1.ResourceRef, error) {
+	_, dynamicClient, discoveryClient, err := s.GetClients(ctx, cluster)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "unable to get the k8s client: '%v'", err)
+	}
+
+	refs := []*corev1.ResourceRef{}
+
+	// fetch every possible kubernetes resource list that is available in the cluster
+	apiResourceLists, err := discoveryClient.ServerPreferredResources()
+	if err != nil {
+		return nil, err
+	}
+
+	// create channel to collect the matching k8s resources thus avoid blocking the waitgroup
+	// ussing a buffered channel with size 100; assumption: 60% native k8s resources + 40 CRDs
+	ch := make(chan resourceAndGvk, 100)
+	wg := sync.WaitGroup{}
+
+	// for each kubernetes resource list, filter out those not being having the verb "list"
+	for _, apiResourceList := range apiResourceLists {
+		if len(apiResourceList.APIResources) == 0 {
+			continue
+		}
+		gv, err := schema.ParseGroupVersion(apiResourceList.GroupVersion)
+		if err != nil {
+			continue
+		}
+
+		// for each kubernetes resource, spwan a goroutine to check if it matches the listOptions
+		for _, resource := range apiResourceList.APIResources {
+			if len(resource.Verbs) == 0 {
+				continue
+			}
+			// filter to resources that support the specified "List" verb
+			if !sets.NewString(resource.Verbs...).HasAll([]string{"list"}...) {
+				continue
+			}
+			wg.Add(1)
+			// spawn a goroutine to fetch the matching k8s resources
+			go func(resource metav1.APIResource, ch chan resourceAndGvk, wg *sync.WaitGroup) {
+				// convert the gvk to a gvr
+				gvk := gv.WithKind(resource.Kind)
+				gvr, _, err := s.kindToResource(s.restMapper, gvk)
+				if err != nil {
+					ch <- resourceAndGvk{nil, err}
+				}
+
+				// ignore the namespace as the resources may appear in different ns from which the PackageInstall is created
+				resources, err := dynamicClient.Resource(gvr).List(ctx, listOptions)
```

Comment:
Interesting, I'd thought the above syntax was for non-namespaced (ie. cluster-level) resources, rather than resources across all namespaces, but see that the dynamic client apparently supports both (`NamespaceableResourceInterface` which is [implemented to do both](https://github.com/kubernetes/client-go/blob/v0.23.1/dynamic/dynamiclister/lister.go))